gklog.1
.TH GKLOG 1 "November 25th" "GNU" "Debian GNU/Linux manual"
.\" NAME should be all caps, SECTION should be 1-8, maybe w/ subsection
.\" other parms are allowed: see man(7), man(1)
.SH NAME
gklog \- monitors and colorizes logs
.SH SYNOPSIS
.B gklog
.I "<log_name> [OPTIONS]"
.SH "DESCRIPTION"
This manual page documents briefly the
.BR gklog
program.
.PP
.B gklog
will read from a log and analyze it, depending on the options the
user passes to it. It reads each line of the log file, compares it
with the patterns at the config file 'till it finds a matching one.
If so, it prints that line with the specified color, else it prints
it with normal color. Use '-' as <log_name> to read from stdin.
.SH OPTIONS
GKLog options are included below.
.TP
.B \-h, \-\-help
Print a help message and exits.
.TP
.B \-c FILENAME, \-\-config FILENAME
Uses an alternate config file, instead of /etc/gklog.rules
.TP
.B \-o FILENAME, \-\-output FILENAME
Outputs to an alternative file, instead of the default 
/var/log/gklog
.TP
.B \-r, \-\-restrict
Does not output non-matching lines. It is usefull for separating
specific lines from the log.
.B \-q, \-\-quiet
Do not output program headers (disclaimer)
.SH "SEE ALSO"
This program will be very usefull when used with tail(1) and
with a pipe (|) together. Read the README file.
.SH AUTHOR
This manual page was written by Gustavo Noronha Silva.


changelog
GKLog 0.7 - 04-12-2000 - 18:19
	* Now gklog looks for configuration on user's $HOME first for 
	  configuration, and if it does not find, looks for the global configuration
	* New option: -q (or --quiet)... It is intended to disallow the output
	  of the also new disclaimer shown when gklog starts.
	* Translated some code from portuguese to english to help everyone
	  on reading the code... 
	* Removed the 'am's and 'pm's of the changelog as it does not make
	  sense showing hour in 24 hours mode and showing am or pm =)
	* Put the manpage made for the Debian distribution on the official
	  source code and edited Makefile to install it.

GKLog 0.6 - 24-11-2000 - 23:57
	* Now you may add blank lines whererver you want and add as many
	  comments as you wish, but you may not include a commented line
	  after the last rule
	* Translated some things in the code that were in portuguese to
	  english, fixed the missing options in --help output

GKLog 0.5 - 09-11-2000 - 20:50
	* Added ways to put comments on the config file (still a little
	   broken, as it is still not possible to put a blank line on it)

GKLog 0.4	- 04-11-2000 - 02:13
	* Added the option -c wich can be used to make gklog to use
	  an alternate configfile other than /etc/gklog.rules

GKLog 0.3	- 28-10-2000 - 11:50
	* Corrected many issues about the rules handling and created three
	  options to use in the command line. Changed the default output
	  from /var/log/firewall to /var/log/gklog and the default log to
	  read from /var/log/messages to /var/log/syslog
	
GKLog 0.2	- 27-10-2000
	* Created new system of user defined rules and colors instead of
	  the built-in ones from 0.1
			
GKLog 0.1	- 26-10-2000
	* Initial Release
	

- Gustavo Noronha Silva (KoV) <dockov@zaz.com.br>


gklog.c
#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#include <errno.h>

#include "read-conf.h"

extern int errno;

// Program's name
#define PrgName "GKLog"
// Program's version
#define PrgVersion "0.7"

// Will the program generate 'header' output?
#define QUIET_DEFAULT 0

// default file to log's output
#define OUTPUT "/var/log/gklog"
// will we show the resulting output to stdout?
#define DEFAULT_SHOW 1

int Show = DEFAULT_SHOW;

/* will the log be shown entirely or will show only mathcing 
lines?*/
int restrict = 0;

/* The location of the config file will be handled in main() but 
this defines where is located the global configuration file */
#define GLOBALCONF "/etc/gklog.rules"
char conf[256];  // Default conffile's name

int ReadLog (char *file, char* Output)
{
    FILE* output; // output file
    FILE* log;
    char Line[1024];
    char c;
    int i;

    if (!strcmp(file, "-"))
    {
    	log = stdin;
    } // check if we're working with stdin
    else
    {
    	// original file
	    log = fopen (file, "r");
     }

    output = fopen (Output, "w");    // file to be created

    if (output == NULL)
    {
		perror("fopen(): output");
		return 1;
    }

    if (log == NULL)
    {
		perror("fopen(): log");
		return 1;
    }

    while (!feof(log))
    {
	bzero (Line, 1024); // fills line with \0's
	c = '\0'; // \0s the char so that there won't be
	// problens in case it is '\n'
	for (i = 0 ; c != '\n' ; i++)
	{
	    c = fgetc(log);
	    if (c == EOF) // if the file is over
		goto theend; // get out
	    Line[i] = c;
	}
	fflush(log);
	Line[i] = '\0';
    fflush(output);

	    if (Show == 1) // show to stdout
	    {
		if ( ProcessRules (Line) == 1 )
		{
			fprintf (stdout, "%s", Line);
    		fprintf (output, "%s", Line);
			fprintf(stdout, "\033[0m");
			fflush (output);
			fflush (stdout);
		}
		else
		{
			if (!restrict)
				fprintf(stdout, "%s", Line);
		}
	    }
    }



theend:
       if (strcmp (file, "-"))
       {
	    if ( (fclose (log)) == EOF )
	    {
		perror("fclose()");
	    }
	}
    if ( (fclose (output)) == EOF )
    {
	perror("fclose()");
    }

    return 0;
}

void Disclaimer (int quiet_run)
{
	if ( quiet_run == 0 )
	{
		printf ("%s version %s\n"
					"This program is Free Software, licensed under the GPL. You probably\n"
					"received a copy of it with this program, if not you may find it on\n"
					"http://www.fsf.org. This program has no warranties! Use it at your\n"
					"own risk!\n\n", PrgName, PrgVersion);
	}
}

void Help (const char *cmd)
{
	Disclaimer ( 0 );
    printf("Usage: %s <log_file> [options]\n\n", cmd);
    printf(
    	"Options:\n"
		"  -h || --help		 this message\n"
		"  -c || --config <file>	 use <file> instead of /etc/gklog.rules\n"
		"  -r || --restrict     	 restricts the output of non-matching lines\n"
		"  -q || --quiet          do not output program headers (disclaimer)\n"
		"  -o || --output <file>  write to <file> instead of /var/log/gklog\n\n"
	   );
    printf("There'll be the following defaults:\n");
    printf("      log:     /var/log/syslog\n");
    printf("      output: /var/log/gklog\n\n");
}

int main(int argc, char *argv[])
{
    int i;
    int quiet_run = QUIET_DEFAULT;
    char Output[1024];
    char logsname[255];
	char *UserConf = calloc (100, sizeof (char));

    if (argc <2)
    {
		Disclaimer (0);
		printf("Usage: %s <log_file> [options]\n\n", argv[0]);
        printf("There'll be the following defaults:\n");
        printf("      log:     /var/log/syslog\n");
        printf("      output: /var/log/gklog\n\n");
    }

    strcpy (logsname, "/var/log/syslog");
    strcpy (Output, OUTPUT);

	for (i = 0 ; i < argc ; i++)
	{
		if (!strcmp(argv[i], "-h") ||
			!strcmp (argv[i], "--help"))
		{
			Help (argv[0]);
			return 0;
		}
		if (!strcmp(argv[i], "-c") ||
			!strcmp (argv[i], "--config"))
		{
			i++;
			bzero (conf, 256);
			strcpy (conf, argv[i]);
		}
		if (!strcmp(argv[i], "-o") ||
			!strcmp (argv[i], "--output"))
		{
			i++;
			strcpy (Output, argv[i]);
		}
		if (!strcmp(argv[i], "-r") ||
			!strcmp (argv[i], "--restrict"))
		{
			restrict = 1;
		}
		if (!strcmp(argv[i], "-q") ||
			!strcmp (argv[i], "--quiet"))
		{
			quiet_run = 1;
		}

	}

	Disclaimer (quiet_run);

	if ( !strcmp (conf, "") )
	{
		UserConf = getenv ("HOME");
		strcat (UserConf, "/.gklog.rules");
	
		if ( fopen (UserConf, "r") != NULL )
		{
			strcpy (conf, UserConf);
			if (quiet_run == 0)
				fprintf (stderr, "Using user configuration: %s...\n", conf);
		}
		else
		{
			strcpy (conf, GLOBALCONF);
			if (quiet_run == 0)
				fprintf (stderr, "User configuration ($HOME/.gklog.rules) not found...\n"
							"Using global configuration: %s...\n", conf);
		}
	}
	else
	{
		if (quiet_run == 0)
			fprintf (stderr, "Using user-defined configuration: %s...\n", conf);	
	}

	if (quiet_run == 0)
		printf ("Starting analysis:\n"); // newline to separate header from output

    if (argc >= 2)
    {
		strcpy (logsname, argv[1]);
	}

	if(!strcmp (logsname, "-"))
	{	
		while (1)
		{
			if ( (ReadLog (logsname, Output)) == 1)
	    		{        	
				printf("Error reading log\n\n");
				return 1;
	  		}
	  	}
	  }
	  else
	  {
		if ( (ReadLog (logsname, Output)) == 1)
		{
		printf("Error reading log\n\n");
		return 1;
		}
	}

    return 0;

}
